File: akka_essentials/src/main/scala/essentials/actor/scala/chapter2/MapReduceApplication.scala
package essentials.actor.scala.chapter2

import java.util.concurrent.TimeUnit

import akka.actor.{ActorSystem, Props}
import akka.pattern.ask
import akka.util.Timeout
import essentials.actor.scala.chapter2.actor.MasterActor

import scala.collection.mutable.ArrayBuffer
import scala.concurrent.Await

sealed trait MapReduceMessage

case class WordCount(word: String, count: Int) extends MapReduceMessage

case class MapData(dataList: ArrayBuffer[WordCount]) extends MapReduceMessage

case class ReduceData(reduceDataMap: Map[String, Int]) extends MapReduceMessage

case class Result() extends MapReduceMessage

object MapReduceApplication {

  def main(args: Array[String]) {
    val system = ActorSystem("MapReduceApp")
    val master = system.actorOf(Props[MasterActor], name = "master")
    implicit val timeout = Timeout(1, TimeUnit.SECONDS)
    master ! "The quick brown fox tried to jump over the lazy dog and  fell on the dog "
    master ! "Dog is man's best friend"
    master ! "Dog and Fox belong to the same family"
    Thread.sleep(1000)
    val future = ask(master, Result).mapTo[String]
    val result = Await.result(future, timeout.duration)
    println(result)
    system.terminate()
  }

}

File: akka_essentials/src/main/java/chapter2java8/MapReduceApplication.java
package chapter2java8;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.pattern.Patterns;
import akka.util.Timeout;

import chapter2.message.Result;
import chapter2java8.actor.MasterActor;
import scala.concurrent.Await;
import scala.concurrent.Future;

import java.util.concurrent.TimeUnit;

public class MapReduceApplication {
    public static void main(String[] args) throws Exception {
        Timeout timeout = new Timeout(1, TimeUnit.SECONDS);
        ActorSystem actorSystem = ActorSystem.create("MapReduceAppJava8");

        ActorRef master = actorSystem.actorOf(Props.create(MasterActor.class, MasterActor::new), "master");
        master.tell("The quick brown fox tried to jump over the lazy dog and fell on the dog", master);
        master.tell("Dog is man's best friend", master);
        master.tell("Dog and Fox jump to the same family", master);
        Thread.sleep(1000);
        Future<Object> future = Patterns.ask(master, new Result(), timeout);
        String result = (String) Await.result(future, timeout.duration());
        System.out.println(result);
        actorSystem.shutdown();
    }
}